be/src/storage/tablet_updates.cpp

```diff
@@ -2620,6 +2623,209 @@ Status TabletUpdates::compaction(MemTracker* mem_tracker) {
+    std::set<PKSizeTieredLevel*, LevelComparator> priority_levels;
+
+    for (auto& [level, candidates] : candidates_by_level) {
+        std::unique_ptr<CompactionInfo> level_info = std::make_unique<CompactionInfo>();
+        level_info->start_version = version;
+        std::sort(candidates.begin(), candidates.end());
+
+        int64_t score = 0;
+        int64_t total_segments = 0;
+        int64_t total_rows = 0;
+        int64_t total_bytes = 0;
+        int64_t total_rows_after_compaction = 0;
+        int64_t total_bytes_after_compaction = 0;
+        for (auto& e : candidates) {
+            size_t new_rows = total_rows_after_compaction + e.num_rows - e.num_dels;
+            size_t new_bytes = total_bytes_after_compaction;
+            if (e.num_rows != 0) {
+                new_bytes += e.bytes * (e.num_rows - e.num_dels) / e.num_rows;
+            }
+            if (total_bytes_after_compaction > 0 && new_bytes > config::update_compaction_result_bytes * 2) {
+                break;
+            }
+            level_info->inputs.push_back(e.rowsetid);
+            score += e.score_per_row * (e.num_rows - e.num_dels);
+            total_rows_after_compaction = new_rows;
+            total_bytes_after_compaction = new_bytes;
+            total_rows += e.num_rows;
+            total_bytes += e.bytes;
+            total_segments += e.num_segments;
+            if (total_bytes_after_compaction > config::update_compaction_result_bytes ||
+                level_info->inputs.size() >= config::max_update_compaction_num_singleton_deltas) {
+                break;
+            }
+        }
+        auto pk_level = std::make_unique<PKSizeTieredLevel>(std::move(level_info), score, level, total_segments,
+                                                            total_rows, total_bytes, total_rows_after_compaction,
+                                                            total_bytes_after_compaction);
+        priority_levels.emplace(pk_level.get());
+        order_levels.emplace_back(std::move(pk_level));
+    }
+
+    size_t version_count = rowsets.size() - info->inputs.size() + _pending_commits.size();
+    // too many rowsets, try to trigger compaction again
+    if (version_count >= config::tablet_max_versions * 80 / 100) {
+        LOG(INFO) << strings::Substitute(
+                "tablet:$0 will try to trigger compaction again because of too many compaction version_count:$1, "
+                "pending:$2",
+                _tablet.tablet_id(), version_count, _pending_commits.size());
+    } else {
+        // give 10s time gitter, so same table's compaction don't start at same time
+        _last_compaction_time_ms = UnixMillis() + rand() % 10000;
+    }
+    PKSizeTieredLevel* selected_level = nullptr;
+    if (!priority_levels.empty()) {
+        selected_level = *priority_levels.begin();
+        if (selected_level->level != -1) {
+            for (auto& e : candidates_by_level[-1]) {
```

Comment:
```suggestion
            for (auto& e : candidates_by_level[selected_level->level]) {
```

Reply:
> 

The rowsets of level `-1` are empty rowsets and they can merge with any other level rowsets because empty rowsets have no disk IO. So we merge all rowsets in level `-1` in every compaction.